contracts/TokenGovernance.sol

```diff
@@ -0,0 +1,149 @@
+import "./Vault.sol";
+import "./TxLib.sol";
+
+contract TokenGovernance {
+
+  uint256 constant PROPOSAL_STAKE = 5000000000000000000000;
+  uint32 constant PROPOSAL_TIME = 604800; // 60 * 60 * 24 * 7 = 7 days
+  address public mvgAddress;
+  IERC20 public leapToken;
+  Bridge public bridge;
+  Vault public vault;
+
+  struct Proposal {
+    address initiator;
+    uint32 openTime;
+    bool finalized;
+    uint256 yesVotes;
+    uint256 noVotes;
+    mapping(address => int256) votes;
+    mapping(bytes32 => address) usedTxns;
+  }
+
+  mapping(bytes32 => Proposal) public proposals;
+
+  event ProposalRegistered(bytes32 indexed proposalHash, address indexed initiator);
+  event VoiceCasted(bytes32 indexed proposalHash, address indexed subject, uint256 weight, bytes32 txHash);
+  event VoiceChallenged(bytes32 indexed proposalHash, address indexed subject, address challenger);
+  event ProposalFinalized(bytes32 indexed proposalHash, bool isApproved);
+
+  constructor(address _mvgAddress, IERC20 _leapToken, Vault _vault) public {
+    mvgAddress = _mvgAddress;
+    leapToken = _leapToken;
+    vault = _vault;
+    bridge = _vault.bridge();
+  }
+
+  function _revertVote(Proposal storage proposal, bytes32 txHash) internal {
+    address signer = proposal.usedTxns[txHash];
+    delete proposal.usedTxns[txHash];
+    int256 votes = proposal.votes[signer];
+    delete proposal.votes[signer];
+    if (votes > 0) {
+      proposal.yesVotes -= uint256(votes);
+    } else {
+      proposal.noVotes -= uint256(votes * -1);
+    }
+  }
+
+  function registerProposal(bytes32 _proposalHash) public {
+    // make sure same proposals hasn't been opened before
+    require(proposals[_proposalHash].openTime == 0, "proposal already exists");
+    // get instance of token contract and pull proposal stake
+    leapToken.transferFrom(msg.sender, address(this), PROPOSAL_STAKE);
+    // create a new proposal in storage
+    proposals[_proposalHash] = Proposal(msg.sender, uint32(now), false, 0, 0);
+    // emit event for frontend
+    emit ProposalRegistered(_proposalHash, msg.sender);
+  }
+
+  function castVote(bytes32 _proposalHash, bytes32[] memory _proof, uint8 _outputIndex, bool isYes) public {
+    Proposal memory proposal = proposals[_proposalHash];
+    require(proposal.openTime > 0, "proposal does not exist");
+    require(proposal.finalized == false, "proposal already finalized");
+    uint32 timestamp;
+    (, timestamp,,) = bridge.periods(_proof[0]);
+    require(timestamp > 0, "The referenced period was not submitted to bridge");
+    require(timestamp <= proposal.openTime, "The transaction was submitted after the vote open time");
+    
+    bytes memory txData;
+    bytes32 txHash;
+    (, txHash, txData) = TxLib.validateProof(96, _proof);
+
+    // parse tx and check if it is usable for voting
+    TxLib.Tx memory tx = TxLib.parseTx(txData);
+    TxLib.Output memory out = tx.outs[_outputIndex];
+    require(out.value > 0, "UTXO has no value");
```

Comment:
I remember that we had a conversation, discussing if the UTXO should be of certain value, not just ANY `Good UTXO`